File: Ginger/src/net/pic4pic/ginger/utils/LogBag.java
package net.pic4pic.ginger.utils;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.TimeZone;
import java.util.UUID;

import net.pic4pic.ginger.entities.LogReportingLevel;
import net.pic4pic.ginger.service.Service;

import com.google.gson.annotations.SerializedName;

import android.util.Log;

public class LogBag {
	
	protected static final int Debug = 1;
	protected static final int Verbose = 2;
	protected static final int Info = 3;
	protected static final int Metric = 4;
	protected static final int Warning = 5;
	protected static final int Error = 6;
	
	public static final String TagLevel = "Level";
	public static final String TagMessage = "Message";
	public static final String TagException = "Exception";
	
	public static final String TagFileName = "File";
	public static final String TagClassName = "Class";
	public static final String TagMethodName = "Method";
	public static final String TagLineNumber = "Line";
	
	public static final String TagAppType = "AppType";
	public static final String TagAppVersion = "AppVersion";
	
	public static final String TagClientId = "ClientId";
	
	public static final String TagObjectType = "ObjectType";
	public static final String TagObjectId = "ObjectId";
	
	public static final String TagParam = "Param";
	
	@SerializedName("Pairs")
	protected ArrayList<LogBagItem> items = new ArrayList<LogBagItem>();
	
	public LogBag add(String tag, String value){
		
		if(tag != null && tag.length() > 0){
			this.items.add(new LogBagItem(tag, value));
		}
		
		return this;
	}
	
	public LogBag add(Throwable ex){		
		
		if(ex != null){
			return this.add(TagException, ex.toString());
		}
		
		return this;
	}
	
	public LogBag addClientId(UUID clientId){
		
		if(clientId == null){
			clientId = new UUID(0, 0); 
		}
		
		return this.add(TagClientId, clientId.toString());
	}
	
	public LogBag addObject(String objectType, UUID objectId){
		
		if(objectId == null){
			objectId = new UUID(0, 0); 
		}
		
		this.add(TagObjectId, objectId.toString());
		return this.add(TagObjectType, objectType);
	}
	
	public LogBag addParam(Object object){		
		return this.add(TagParam, (object == null ? null : object.toString()));
	}
	
	public void e(){
		this.e(null);
	}
	
	public void e(String message){
		this.log(Error, message);
	}
	
	public void w(){
		this.w(null);
	}
	
	public void w(String message){		
		this.log(Warning, message);
	}
	
	public void m(){		
		this.m(null);
	}
	
	public void m(String message){		
		this.log(Metric, message);
	}
	
	public void i(){
		this.i(null);
	}
	
	public void i(String message){
		this.log(Info, message);
	}
	
	public void v(){
		this.v(null);
	}
	
	public void v(String message){
		this.log(Verbose, message);
	}
	
	public void d (){
		this.d(null);
	}
	
	public void d (String message){
		this.log(Verbose, message);
	}
	
	protected String getLevelAsText(int level){
		
		switch(level){
		
			case Error : 
				return "Error";
				
			case Warning : 
				return "Warning";
			
			case Metric : 
				return "Metric";
				
			case Info : 
				return "Info";
				
			case Debug : 
				return "Debug";
				
			default:
				return "Verbose";
		}
	}
	
	@Override
	public String toString(){
		
		if(this.items.size() == 0){
			return "";
		}
		
		boolean appended = false;
		StringBuilder builder = new StringBuilder();		
		LogBagItem exceptionLogBagItem = null;		
		for(int x=this.items.size()-1; x>=0; x--){
			
			LogBagItem logBagItem = this.items.get(x);
			if(logBagItem.getName().equalsIgnoreCase(TagException)){
				// this will be inserted at the end
				exceptionLogBagItem = logBagItem;
			}
			else{
				
				if(appended){
					builder.append("[;]");
				}
				
				builder.append("[" + logBagItem.toString() + "]");
				appended = true;
			}
		}
		
		if(exceptionLogBagItem != null){
			
			if(appended){
				builder.append("[;]");
			}
			
			builder.append("[" + exceptionLogBagItem.toString() + "]");
			appended = true;
		}
		
		return builder.toString();
	}	
	
	protected String getValueOf(String tag){
		for(LogBagItem item : this.items){
			if(item.getName().equalsIgnoreCase(tag)){
				return item.getValue();
			}
		}
		
		return "";
	}
	
	protected void log(int level, String message){		
		
		// add level
		this.add(TagLevel, this.getLevelAsText(level));
		
		// add message
		if(message != null && message.length() > 0){
			this.add(TagMessage, message);
		}
		
		// log to console
		this.logToConsole(level, false);
		
		// add extra since this needs to be sent to server
		if(this.shouldBeSent(level)){
			
			UUID logKey = UUID.randomUUID();
			this.add("LogId", logKey.toString());		
			SimpleDateFormat dateFormatter = new SimpleDateFormat("MM/dd/yyyy HH:mm:ss");
			dateFormatter.setTimeZone(TimeZone.getTimeZone("GMT"));
			this.add("LogTimeUTC", dateFormatter.format(new Date()));
			
			// send
			LogPusher.instance().add(logKey, this);	
		}
		
		// do not clear! we don't know what happens.
	}
	
	protected boolean shouldBeSent(int level)
	{
		int reportingLevel = Service.getInstance().getLogReportingLevel();
		if(level == LogBag.Error && reportingLevel >= LogReportingLevel.OnlyErrors){
			return true;
		}
		if(level == LogBag.Warning && reportingLevel >= LogReportingLevel.WarningsAndAbove){
			return true;
		}
		if(level == LogBag.Metric && reportingLevel >= LogReportingLevel.MetricsAndAbove){
			return true;
		}
		if(level == LogBag.Info && reportingLevel >= LogReportingLevel.InfoAndAbove){
			return true;
		}
		if(level == LogBag.Verbose && reportingLevel >= LogReportingLevel.VerboseAndAbove){
			return true;
		}
		if(level == LogBag.Debug && reportingLevel >= LogReportingLevel.All){
			return true;
		}
		
		return false;
	}
	
	protected void logToConsole(int level, boolean summarize){
	
		String tag = "Ginger";
		String cls = this.getValueOf(TagClassName);
		if(cls != null && cls.length() > 0){
			tag += "-" + cls;
		}		
		
		String log = "";
		if(summarize){
			String msg = this.getValueOf(TagMessage);
			String exc = this.getValueOf(TagException);
			log = msg == null ? "" : msg;
			if(exc != null && exc.length() > 0){
				if(log.length() > 0){
					log += " : ";
				}
				log += exc;
			}
		}

		if(log.length() <=0 ){
			log = this.toString();
		}
		
		switch(level){
		
			case Error : 
				Log.e(tag, log);
				break;
				
			case Warning : 
				Log.w(tag, log);
				break;
				
			case Metric : 
				Log.i(tag, log); // it is still info
				break;
				
			case Info : 
				Log.i(tag, log);
				break;			
				
			case Debug : 
				Log.d(tag, log);
				break;
				
			default:
				Log.v(tag, log);
				break;
		}
	}
}

File: Pic4Pic/Pic4Pic.ObjectModel/Entities/LogBag.cs
using System;
using System.Collections.Generic;
using System.Runtime.Serialization;
using System.Text;

namespace Pic4Pic.ObjectModel
{
    [DataContract()]
    public class LogProperty : IVerifiable
    {
        [DataMember()]
        public string Name { get; set; }

        [DataMember()]
        public string Value { get; set; }

        public LogProperty()
        {
        }

        public LogProperty(string name, string value)
        {
            this.Name = name;
            this.Value = value;
        }

        public void Validate()
        {
            if (String.IsNullOrWhiteSpace(this.Name))
            {
                throw new Pic4PicArgumentException("Name of LogProperty may not be empty", "Name");
            }
        }

        public override string ToString()
        {
            if (String.IsNullOrWhiteSpace(this.Name))
            {
                return String.Empty;
            }

            if (String.IsNullOrWhiteSpace(this.Value))
            {
                return Name;
            }

            return this.Name + "=" + this.Value;
        }
    }

    [DataContract()]
    public class LogBag : IVerifiable
    {
        private const int Debug = 1;
	    private const int Verbose = 2;
	    private const int Info = 3;
	    private const int Metric = 4;
	    private const int Warning = 5;
	    private const int Error = 6;

        public const String TagAppType = "AppType";
        public const String TagFileName = "File";
        public const String TagClassName = "Class";
        public const String TagMethodName = "Method";
        public const String TagLineNumber = "Line";
        public const String TagLogId = "LogId";
        public const String TagLogTime = "LogTimeUTC";
        private const String TagMessage = "Message";
        private const String TagException = "Exception";
        private const String TagObjectType = "ObjectType";
        private const String TagObjectId = "ObjectId";
        private const String TagParam = "Param";
        private const String TagLevel = "Level";        

        [DataMember()]
        public List<LogProperty> Pairs
        {
            get
            {
                return this.pairs;
            }
            set
            {
                if (value != null)
                {
                    this.pairs = value;
                }
                else
                {
                    this.pairs = new List<LogProperty>();
                }
            }
        }

        private List<LogProperty> pairs = new List<LogProperty>();

        public void Validate()
        {
            if (this.pairs.Count <= 0)
            {
                throw new Pic4PicArgumentException("Property bags in the LogBag may not be empty", "Pairs");
            }

            foreach (LogProperty pair in this.pairs)
            {
                pair.Validate();
            }
        }

        public String GetErrorLevel()
        {
            foreach (LogProperty pair in this.pairs)
            {
                if (pair.Name.ToLower().Equals(TagLevel.ToLower()))
                {
                    return this.convertClientErrorLevelTo(pair.Value);
                }
            }

            return LogCategory.Verbose;
        }

        public String GetByTag(String tagName)
        {
            foreach (LogProperty pair in this.pairs)
            {
                if (pair.Name.ToLower().Equals(tagName.ToLower()))
                {
                    return pair.Value;
                }
            }

            return null;
        }

        public Guid GetGuidByTag(String tagName)
        {
            String val = this.GetByTag(tagName);
            if (String.IsNullOrWhiteSpace(val))
            {
                return Guid.Empty;
            }

            Guid guidVal = Guid.Empty;
            Guid.TryParse(val, out guidVal);
            return guidVal;
        }

        public int GetIntByTag(String tagName, int defaultVal)
        {
            String val = this.GetByTag(tagName);
            if (String.IsNullOrWhiteSpace(val))
            {
                return defaultVal;
            }

            Int32 intVal = defaultVal;
            Int32.TryParse(val, out intVal);
            return intVal;
        }

        public DateTime GetDateTimeByTag(String tagName)
        {
            String val = this.GetByTag(tagName);
            if (String.IsNullOrWhiteSpace(val))
            {
                return default(DateTime);
            }

            DateTime dateTimeVal = default(DateTime);
            try
            {
                dateTimeVal = DateTime.ParseExact(val, "MM/dd/yyyy HH:mm:ss", System.Globalization.CultureInfo.InvariantCulture);
            }
            catch (Exception ex)
            {
                Logger.bag().Add(ex).LogError("Parsing date time has failed.");
            }
            
            return dateTimeVal;
        }

        private String convertClientErrorLevelTo(String clientErrorLevel)
        {
            switch (clientErrorLevel.ToLower())
            {
                case "error":
                    return LogCategory.Error;

                case "warning":
                    return LogCategory.Warning;

                case "metric":
                    return LogCategory.Metric;

                case "info":
                    return LogCategory.Info;

                case "debug":
                    return LogCategory.Debug;

                default:
                    return LogCategory.Verbose;
            }
        }

        public override string ToString()
        {
            if (this.pairs.Count <= 0)
            {
                return String.Empty;
            }

            bool appended = false;
            StringBuilder builder = new StringBuilder();

            LogProperty exceptionLogBagItem = null;
            for (int x = this.pairs.Count - 1; x >= 0; x--)
            {
                LogProperty logBagItem = this.pairs[x];
                if (logBagItem.Name.Equals(TagException))
                {
                    // this will be inserted at the end
                    exceptionLogBagItem = logBagItem;
                }
                else
                {
                    if (appended)
                    {
                        builder.Append("[;]");
                    }

                    builder.Append("[" + logBagItem.ToString() + "]");
                    appended = true;
                }
            }

            if (exceptionLogBagItem != null)
            {
                if (appended)
                {
                    builder.Append("[;]");
                }

                builder.Append("[" + exceptionLogBagItem.ToString() + "]");
                appended = true;
            }

            return builder.ToString();
        }

        public LogBag Add(String tag, String value)
        {

            if (!String.IsNullOrWhiteSpace(tag))
            {
                this.pairs.Add(new LogProperty(tag, value));
            }

            return this;
        }

        public LogBag Add(Exception ex)
        {

            if (ex != null)
            {
                return this.Add(TagException, ex.ToString());
            }

            return this;
        }

        public LogBag AddObject(String objectType, Guid objectId)
        {
            this.Add(TagObjectId, objectId.ToString());
            return this.Add(TagObjectType, objectType);
        }

        public LogBag AddParam(Object obj)
        {
            return this.Add(TagParam, (obj == null ? null : obj.ToString()));
        }

        public void LogError()
        {
            this.LogError(null);
        }

        public void LogError(String message)
        {
            this.log(Error, message);
        }

        public void LogWarning()
        {
            this.LogWarning(null);
        }

        public void LogWarning(String message)
        {
            this.log(Warning, message);
        }

        public void LogMetric()
        {
            this.LogMetric(null);
        }

        public void LogMetric(String message)
        {
            this.log(Metric, message);
        }

        public void LogInfo()
        {
            this.LogInfo(null);
        }

        public void LogInfo(String message)
        {
            this.log(Info, message);
        }

        public void LogVerbose()
        {
            this.LogVerbose(null);
        }

        public void LogVerbose(String message)
        {
            this.log(Verbose, message);
        }

        public void LogDebug()
        {
            this.LogDebug(null);
        }

        public void LogDebug(String message)
        {
            this.log(Verbose, message);
        }

        protected String getLevelAsText(int level)
        {
            switch (level)
            {
                case Error:
                    return "Error";

                case Warning:
                    return "Warning";

                case Metric:
                    return "Metric";

                case Info:
                    return "Info";

                case Debug:
                    return "Debug";

                default:
                    return "Verbose";
            }
        }
        
        protected void log(int level, String message)
        {

            // add level
            this.Add(TagLevel, this.getLevelAsText(level));

            // add message
            if (!String.IsNullOrWhiteSpace(message))
            {
                this.Add(TagMessage, message);
            }

            String log = this.ToString();
            System.Diagnostics.Trace.WriteLine(log, "P4P_" + this.getLevelAsText(level));
        }

        public static LogBag Parse(String logLine)
        {
            if (String.IsNullOrWhiteSpace(logLine)) 
            {
                return null;
            }

            if (!logLine.StartsWith("P4P_"))
            {
                return null;
            }

            logLine = logLine.Substring("P4P_".Length);
            int index = logLine.IndexOf(":");
            if(index <= 0)
            {
                return null;
            }
            
            string logType = logLine.Substring(0, index).Trim();
            if (logType.Length == 0)
            {
                return null;
            }
            
            logLine = logLine.Substring(index).TrimStart(' ', ':');
            
            // split
            String[] tokens = logLine.Split(new string[] {"[;]"}, StringSplitOptions.RemoveEmptyEntries);
            if (tokens.Length == 0)
            {
                return null;
            }

            Dictionary<String, String> pairs = new Dictionary<string, string>();
            pairs.Add(LogBag.TagLevel, logType);

            foreach (String token in tokens)
            {
                if (token.Trim().Length > 0)
                {
                    String temp = token.Trim();
                    index = temp.IndexOf("=");
                    if (index > 0 && index < (temp.Length-1))
                    {
                        String key = temp.Substring(0, index).Trim(' ', '[', ']');
                        String val = temp.Substring(index + 1).Trim(' ', '[', ']');

                        if (key.Length > 0 && val.Length > 0)
                        {
                            if (pairs.ContainsKey(key))
                            {
                                pairs[key] = val;
                            }
                            else
                            {
                                pairs.Add(key, val);
                            }
                        }
                    }
                }
            }

            if (pairs.Count <= 1)
            {
                return null;
            }

            LogBag bag = new LogBag();
            foreach (String key in pairs.Keys)
            {
                bag.Pairs.Add(new LogProperty(key, pairs[key]));
            }

            return bag;
        }
    }
}
